extensions/2.0/Khronos/KHR_accessor_float64/README.md

```diff
@@ -0,0 +1,129 @@
+<!--
+Copyright 2024 The Khronos Group Inc.
+SPDX-License-Identifier: LicenseRef-KhronosSpecCopyright
+-->
+
+# KHR\_accessor\_float64
+
+## Contributors
+
+- Alexey Knyazev, [@lexaknyazev](https://github.com/lexaknyazev)
+- Sean, [@spnda](https://github.com/spnda)
+
+Copyright 2024 The Khronos Group Inc.
+See [Appendix](#appendix-full-khronos-copyright-statement) for full Khronos Copyright Statement.
+
+## Status
+
+Draft
+
+## Dependencies
+
+Written against the glTF 2.0 Specification.
+
+## Overview
+
+The base glTF 2.0 Specification supports only one floating-point component type - IEEE-754 _binary32_ (also known as single-precision). Although this is sufficient for most applications, there are use cases that benefit from increased range and precision provided by the 64-bit floating-point data type (also known as double-precision).
+
+This extension expands the set of recognized accessor component types with a new enum value `5130` that represents IEEE-754 _binary64_.
+
+This extension allows usage of such accessors in all places where the single-precision floating-point component type is accepted by the base glTF 2.0 Specification or extensions.
+
+Because this extension does not provide a way to specify alternative data representations, assets that use this extension **MUST** specify it in the `extensionsRequired` array.
+
+## Extending Accessors
+
+When the extension is supported, the following data type is valid for accessors in addition to the types defined in [Section 3.6.2.2](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#accessor-data-types).
+
+| `componentType` | Data Type | Signed | Bits |
+|-----------------|-----------|--------|------|
+| `5130`          | _double_  | Signed | 64   |
+
+Such accessors **MUST** adhere to the following additional rules:
+
+- their `normalized` property **MUST NOT** be set to true;
+- their effective byte offsets **MUST** be aligned to 8-byte boundaries;
+- when referring to data in a GLB-stored buffer, the start of the buffer **MUST** be aligned to 8-byte boundaries by adding necessary padding bytes to the JSON chunk as defined in [Section 4.4.3.2](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#structured-json-content).
```

Comment:
Is it really necessary to require that alignment for the _buffer_? 

I thought that only the start of the accessor (accessor.byteOffset + bufferView.byteOffset) needs to be aligned accordingly, as of https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#data-alignment . As it is written now, it will have implications on the 'chunk' alignment of https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#chunks-overview 

(Maybe I need a refresher for some implications, for things like the JavaScript `Float64Array`, and what _exactly_ the alignment requirements refer to in this case. But I thought that having a `Float64Array` with an offset of 16 inside an `ArrayBuffer` would be OK, even if the glTF buffer formally starts at offset 12 of this `ArrayBuffer`, for example)

Reply:
If the whole GLB is received as an `ArrayBuffer`, apps won't be able to create a `Float64Array` from it without the 8-byte buffer alignment requirement.

```js
// Uncaught RangeError: start offset of Float64Array should be a multiple of 8
new Float64Array(new ArrayBuffer(16), 4)
```

Without this requirement, apps would have to make an extra copy. This sounds worse than adding four trailing spaces to JSON during export if needed.

Reply:
After revisiting this one more time I think that this requirement could be demoted to suggestion, i.e., saying SHOULD instead of MUST.

The reasoning is that some pipeline tools that produce GLB files may be able to passthrough assets using this extension but won't be updated any time soon to specifically handle the 8-byte alignment.

Reply:
My thoughts had been about the point is that this `Float64Array` will not be created for the whole buffer, but only for the _accessor_ that is _part_ of the buffer. So...
```
accessor = new Float64Array(glbArrayBuffer, 4 + 4)
                                            ^   ^offset of BV/accessor in buffer
                                            |offset of BIN buffer in GLB 
```
would be valid. 

But I can see that the alignment requirement has to be phrased in terms of the `buffer` (i.e. ~"the `BIN` chunk"). That's the only way to ensure that 'everything works', regardless of whether _the whole GLB_ is stored in one `ArrayBuffer`, or whether only the BIN chunk is stored in one `ArrayBuffer`.

---

I think it could be tricky to support this for _exporters_: Usually, they'll do 
```
glbData.fillWith(jsonData);
glbData.padTo(4);
glbData.fillWith(bufferData);
```
But now they'll have to do
```
glbData.fillWith(jsonData);
if (bufferData.containsAccessorWith64bitValues()) {
    glbData.padTo(8);
} else {
    glbData.padTo(4);
}
glbData.fillWith(bufferData);
```
where that `containsAccessorWith64bitValues` could be difficult to implement. 

Exporters might be tempted to just use the "lazy" solution of _always_ aligning to 8, to be safe. But that might raise questions in view of https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#glb-stored-buffer which says

> The byte length of the BIN chunk MAY be up to 3 bytes bigger than JSON-defined buffer.byteLength value to satisfy GLB padding requirements.

and where the BIN chunk may become up to 7 bytes bigger than the `buffer.byteLength` when this "lazy" approach is used. I think the validator would then complain about `BUFFER_GLB_CHUNK_TOO_BIG`.

(Note: Nothing of this is "critical". I'm just trying to think through corner cases and implications...)

Reply:
The BIN chunk trailing bytes are unrelated to accessor alignment. They are needed only to ensure 4-byte alignment of hypothetical custom chunks present in GLB afterwards.

Reply:
Exporters will have to ensure the 8-byte alignment of the start of the `BIN` chunk (by padding the JSON chunk). Strictly speaking, they _only_ have to do this when the glTF uses this extension. But I think that they will _always_ do it. And that's OK. 

(They'll probably figure out that they should _not_ align _all_ chunks starts/ends to 8 bytes (ignoring what the alignment requirements for a 'custom' chunk could be), so maybe that's not so much of an issue...)

Reply:
Relaxed the GLB-related language to not invalidate all existing tools that passthrough unknown extensions.

Please review the rest of the extension text.

Reply:
The remaining text looks fine for me (I'd probably need more time (and maybe try to _implement_ it...) to find _potential_ pitfalls, but the description itself _looks_ straightforward - let's see whether others have further comments here)